File: services/products/src/main/java/br/com/deltaglobalbank/products/domain/product/ProductRepository.java
package br.com.deltaglobalbank.products.domain.product;

import java.util.List;
import java.util.UUID;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductRepository {
    Product save(Product product);

    Product findById(UUID id, UUID tenantId);

    boolean existsActiveByTypeAndAgreement(UUID tenantId, ProductType type, String agreementName);

    Page<Product> findLendingPage(UUID tenantId, String agreementName, Boolean active, Pageable pageable);

    boolean existsAnotherActiveWithAgreement(UUID tenantId, ProductType type, String agreementName, UUID excludeId);

    List<Product> findProducts(UUID tenantId, ProductType type, Boolean active, String agreementName);
}

File: services/products/src/main/kotlin/br/com/deltaglobalbank/products/domain/product/ProductRepository.kt
package br.com.deltaglobalbank.products.domain.product

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import java.util.UUID

interface ProductRepository {
    fun save(product: Product): Product
    fun findById(id: UUID, tenantId: UUID): Product?
    fun existsActiveByTypeAndAgreement(tenantId: UUID, type: ProductType, agreementName: String): Boolean
    fun findLendingPage(tenantId: UUID, agreementName: String?, active: Boolean?, pageable: Pageable): Page<Product>
    fun existsAnotherActiveWithAgreement(tenantId: UUID, type: ProductType, agreementName: String, excludeId: UUID): Boolean
    fun findProducts(tenantId: UUID, type: ProductType?, active: Boolean?, agreementName: String?): List<Product>
}